Add distinguished select game screen

BdeBock opened the pull request:
fixes #2173 
Add distinguished select game screen for singleplayer and host game.
Singleplayer now has a subtitle "singleplayer" and host game has a subtitle "multiplayer".
Created a SelectGameScreenHost class (copy of SelectGameScreen) just to be able to couple the right .ui file to it (no idea if this is the right way to make the distinction)

GooeyHub commented:
Can one of the admins please verify this patch?

msteiger commented:
Hi Bas - what's wrong?

BdeBock commented:
Hi Martin, I fixed the issue, but wanted my team ( @AlexanderGrooff , @JanZegers , @arkka ) to go over it first before pull requesting it to you guys. Though as you can see I accidentally created the PR already, that's why I closed it again.

GooeyHub commented:
Can one of the admins please verify this patch?

Cervator commented:
ok to test

Hi @BdeBock and thanks for the PR :-)

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/440/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Thanks for your contribution, Bas. I'd prefer to use one class for both cases, since they are highly similar, otherwise this results in [code duplication](https://en.wikipedia.org/wiki/Duplicate_code#Problems_associated_with_duplicate_code). You can still set different subtitles, if that's the only difference between the two.

In fact, the class already differentiates between the two modes in the action listener for the "Start Game" button (the `isLoadingAsServer` flag controls the behavior).

You should be able to rename/reuse the flag to respect the subtitle id as well.

AlexanderGrooff commented:
It seems that the SelectGameScreenHost.java was never reached, meaning that it's effectively dead code.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/442/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Thanks!
The `selectGameScreenHost` UI code looks like a duplicate as well.
Also, `mainmenu.skin/subtitle` could be redundant as it is probably defined in `default.skin` already.

Cervator commented:
Bump - should we merge this or are we still waiting for a caption tweak instead of unique new `.ui` file?

msteiger commented:
I think that only 5-6 lines of code need to be changed so solve the issue, so I'd either wait until this PR is updated or create a new fix.

BdeBock commented:
Kind of busy at the moment, will fix it in the weekend.

Cervator commented:
No problem, just wanted to check :-)

BdeBock commented:
Hi guys, looked at it again. 
Martin mentioned selectGameScreenHost.ui looks a lot like selectGameScreen.ui. This is true except for the difference of the subtitle label. But I see no other way to let the selectGameScreen of singleplayer and multiplayer have a different sub-heading, without letting them have their own .ui file. 
For the second point Martin mentioned, the mainmenu.skin/subtitle is defined as there wasn't one yet for a subtitle/sub-heading. And I believe it should be placed in the mainmenu.skin as it is part of the mainmenu. 

Could be I'm totally wrong about this, if so please let me know ;p

Cervator commented:
Hey again @BdeBock thanks for following up :-)

I figure there's some way to use a variable in the .ui file that in turn will point to an i18n string key. I just don't know how.

Hopefully @msteiger will see this tomorrow and know of an answer :-)

msteiger commented:
You can use a placeholder text for the subtitle label and set the correct one through Java code when the menu screen is opened.

This is done at several occasions, for example in the module selection dialog, which updates the module description text according to the module selection.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/479/
Hooray Jenkins reported success with all tests good!

BdeBock commented:
Implemented the proposed solution of msteiger, think everything is ok now.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/480/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/481/
Hooray Jenkins reported success with all tests good!

msteiger commented on engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/MainMenuScreen.java:
```diff
@@ -34,7 +34,12 @@
         find("version", UILabel.class).setText(TerasologyVersion.getInstance().getHumanVersion());
-        WidgetUtil.trySubscribe(this, "singleplayer", button -> getManager().pushScreen("engine:selectGameScreen"));
+        WidgetUtil.trySubscribe(this, "singleplayer", button -> {
+            UIScreenLayer screen = getManager().pushScreen("engine:selectGameScreen");
```
There should be a similar method with second parameter that takes a class argument. It returns an arbitrary type which makes the check below unnecessary.

msteiger replied:
The exact method call is: [NUIManager.pushScreen(String screenUri, Class<T> expectedType);](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/rendering/nui/NUIManager.java#L66). It returns a `SelectGameScreen` you can use directly.

msteiger commented on engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/SelectGameScreen.java:
```diff
@@ -47,6 +49,20 @@
     @Override
     public void initialise() {
+
+        UILabel subtitle2 = find("subtitle2", UILabel.class);
```
Good! Can you find a better name than `subtitle2`?

msteiger commented on engine/src/main/resources/assets/skins/mainmenu.skin:
```diff
@@ -35,6 +35,10 @@
             "font": "NotoSans-Regular-Large",
             "text-color": "F0F0F0FF"
         },
+         "subtitle": {
```
Agreed!

msteiger commented on engine/src/main/resources/assets/ui/menu/selectGameScreenHost.ui:
```diff
@@ -0,0 +1,119 @@
+{
```
Is this a duplicate?

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/484/
Hooray Jenkins reported success with all tests good!

BdeBock commented:
Changed the name of subtitle2 to gameTypeTitle. Changed the static multi- and singleplayer strings to the translationSystem. Removed the duplicate selectGameScreenHost.ui.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/485/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Thank you, Bas! Only one open issue to address and I will be happy to merge your contribution.

BdeBock commented:
Team member is reviewing my solution for that issue, will be fixed tomorrow.

BdeBock commented:
Last issue should be fixed now.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/487/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/488/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/489/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Great - thank you for being so persistent.

BdeBock commented:
No problem, thanks for the help.

Cervator commented:
Much appreciated all :-)

Also I feel terrible saying this, but we have the new subtitle now on the single player and multiplayer screens - but if you then go to "Create" game in either case you are now again on an anonymous screen without the new subtitle... :D

Up for adding the subtitle on the create screen as well? On the plus side it should be a whole lot easier this time :-)
